File: astropix3rowcol.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.colors
import argparse

def mkplots(m, m2=None):
  # Plot 1
  plt.hist(m.row_timestamp-m.col_timestamp, bins=np.arange(-10, 11, 1))
  if m2 is not None: plt.hist(m2.row_timestamp-m2.col_timestamp, bins=np.arange(-10, 11, 1))
  plt.yscale("log")
  plt.grid(True, "both")
  plt.xlabel("Row TS - Col TS")
  plt.ylabel("Matched hits")
  plt.grid(True, "both"); plt.show()
  # Plot 2
  plt.hist(m.row_tot-m.col_tot, bins=np.arange(-50, 50, 1))
  if m2 is not None: plt.hist(m2.row_tot-m2.col_tot, bins=np.arange(-50, 50, 1))
  plt.yscale("log")
  plt.grid(True, "both")
  plt.xlabel("Row ToT - Col ToT")
  plt.ylabel("Matched hits")
  plt.grid(True, "both"); plt.show()
  # Plot 3
  plt.hist(m.row_fpga_ts-m.col_fpga_ts, bins=np.arange(-500, 50, 1))
  if m2 is not None: plt.hist(m2.row_fpga_ts-m2.col_fpga_ts, bins=np.arange(-500, 50, 1))
  plt.yscale("log")
  plt.grid(True, "both")
  plt.xlabel("Row FPGA TS - Col FPGA TS")
  plt.ylabel("Matched hits")
  plt.grid(True, "both"); plt.show()
  # Plot 4
  plt.scatter(m.row_tot, m.row_tot-m.col_tot, s=.1)
  if m2 is not None: plt.scatter(m2.row_tot, m2.row_tot-m2.col_tot, s=.1)
  plt.xlabel("Row ToT")
  plt.ylabel("Row ToT - Col ToT")
  plt.grid(True, "both"); plt.show()
  # Plot 5
  if m2 is None: plt.hist2d(m.row_timestamp - m.col_timestamp, m.row_tot-m.col_tot, (np.arange(-4, 6, 1), np.arange(-4000, 4300, 10)), norm=matplotlib.colors.LogNorm())
  else: plt.hist2d(m2.row_timestamp - m2.col_timestamp, m2.row_tot-m2.col_tot, (np.arange(-4, 6, 1), np.arange(-4000, 4300, 10)), norm=matplotlib.colors.LogNorm())
  plt.colorbar()
  plt.xlabel("Row TS - Col TS")
  plt.ylabel("Row ToT - Col ToT")
  plt.grid(True, "both"); plt.show()

  

def rowcolmatch(chip0, fts=lambda x,y: x-y==0 or x-y==1, ftot=lambda x,y: x-y>6 and x-y<15):
  """
  Performs Row-Col matching for AstroPix V3 (A-STEP firmware)
  :param chip0: pandas DataFrame of decoded hits
  :param fts: function (int, int) -> bool, condition on the row and col TS to match half-hits
  :param ftot: function (int, int) -> bool, condition on the row and col ToT_total value to match hits (default: True)
  :retuns: pandas DataFrame of matched hits
  """
  #outdict = {"dec_ord":[], "readout":[], "layer":[], "chipID":[], "payload":[], "location":[], "isCol":[], "timestamp":[], "tot_msb":[], "tot_lsb":[], "tot_total":[], "tot_us":[], "fpga_ts":[]}
  outdict = {"layer":[], "chipID":[], "row":[], "col":[], "row_timestamp":[], "col_timestamp":[], 
              "row_tot":[], "col_tot":[], "row_tot_us":[], "col_tot_us":[], "row_fpga_ts":[], "col_fpga_ts":[]}
  linenb = 0
  while linenb < len(chip0):
    index = chip0.index[linenb]
    #print(f"line={linenb}, i={index}, isCol={chip0.isCol[index]}, ts={chip0.timestamp[index]}")
    if chip0.isCol[index] == 0: # Try to match this row to a col
      foundcol = False
      i = linenb + 1
      while i < len(chip0) and (not(foundcol) or chip0.isCol[chip0.index[i]]==1):
        colindex = chip0.index[i]
        if chip0.isCol[colindex]==1: # Skip other rows
          if not(foundcol): foundcol=True
          #print(chip0.timestamp[index], chip0.index[i], chip0.isCol[chip0.index[i]], chip0.timestamp[chip0.index[i]])
          if fts(chip0.timestamp[index], chip0.timestamp[colindex]) and ftot(chip0.tot_total[index], chip0.tot_total[colindex]):
            #print(f"{index}\t{chip0.index[i]}\t{chip0.timestamp[index]}\t{chip0.timestamp[chip0.index[i]]}\t{chip0.tot_total[index]}\t{chip0.tot_total[chip0.index[i]]}")
            outdict["layer"].append(chip0.layer[index])
            outdict["chipID"].append(chip0.chipID[index])
            outdict["row"].append(chip0.location[index])
            outdict["col"].append(chip0.location[colindex])
            outdict["row_timestamp"].append(chip0.timestamp[index])
            outdict["col_timestamp"].append(chip0.timestamp[colindex])
            outdict["row_tot"].append(chip0.tot_total[index])
            outdict["col_tot"].append(chip0.tot_total[colindex])
            outdict["row_tot_us"].append(chip0.tot_us[index])
            outdict["col_tot_us"].append(chip0.tot_us[colindex])
            outdict["row_fpga_ts"].append(chip0.fpga_ts[index])
            outdict["col_fpga_ts"].append(chip0.fpga_ts[colindex])
        i += 1
    linenb += 1 # Skip cols
  return pd.DataFrame.from_dict(outdict)

if __name__ ==  "__main__":
  parser = argparse.ArgumentParser("Row-Column matching script for AstroPix v3")
  parser.add_argument("filename", help="Name of the decoded .csv file")
  parser.add_argument("-q", "--quiet", action="store_true", help="Suppresses output")
  parser.add_argument("-l", "--layers", type=int, default=3, help="Number of layers/lanes, default=3")
  parser.add_argument("-c", "--chips", type=int, default=4, help="Number of chips per layer/lane, default=4")
  parser.add_argument("--mints", type=int, default=0, help="Minimum halfhit Timestamp difference (row-col) to match, default=0")
  parser.add_argument("--maxts", type=int, default=1, help="Minimum halfhit Timestamp difference (row-col) to match, default=1")
  parser.add_argument("--mintot", type=int, default=6, help="Minimum halfhit ToT difference (row-col) to match, default=7")
  parser.add_argument("--maxtot", type=int, default=15, help="Minimum halfhit ToT difference (row-col) to match, default=14")
  args = parser.parse_args()
  # Read data
  data = pd.read_csv(args.filename)
  dataf = data[(data.payload==4)&(data.location<35)] # Filter corrupted data
  if not args.quiet:
    print(f"{len(data)} decoded halfhits read, {len(dataf)} halfhits are valid ({100*len(dataf)/len(data):.2f}%).")
  # Row-col match per chip
  output = []
  for layer in range(args.layers):
    for chip in range(args.chips):
      datac = dataf[(dataf.chipID==chip)&(dataf.layer==layer)]
      output.append(rowcolmatch(datac, fts=lambda x,y: x-y>=args.mints and x-y<=args.maxts, ftot=lambda x,y: x-y>=args.mintot and x-y<=args.maxtot))
      if not args.quiet:
        print(f"Layer {layer}, Chip {chip}: {len(datac)} halfhits found, {len(output[-1])} hits matched ({100*len(output[-1])*2/max(1, len(datac)):.2f}%).")
  pd.concat(output).to_csv(f"{args.filename[:-4]}_matched.csv")



if __name__ == "__main__" and False: # Set to True to run as a script
  print("Loading data")
  d = pd.read_csv("/Users/alaviron/AstroPix/data/20250624-172917.csv")
  d2 = d[(d.payload==4)&(d.location<35)] # Filter rubbish
  print(f"{100*(1-len(d2)/len(d))}% of the data has been removed as rubbish")
  print("Select Chip ID = 0")
  chip0 = d2[(d2.chipID==0)&(d2.layer==1)]
  print("Perform row-col matching: Only selection is within 5 TS units and row matched within the next col delivery")
  m = rowcolmatch(chip0, fts=lambda x,y: abs(x-y)<5, ftot=lambda x,y: True)
  mkplots(m)
  print("Perform row-col matching: Delta TS = 0 or 1")
  m2 = rowcolmatch(chip0, lambda x,y: x-y==0 or x-y==1, ftot=lambda x,y: True)
  mkplots(m, m2)
  print("Perform row-col matching: Delta TS = 0 or 1, -10 < Delta ToT < 30")
  m3 = rowcolmatch(chip0, lambda x,y: x-y==0 or x-y==1, lambda x,y: x-y>6 and x-y<15)
  mkplots(m, m3)

File: rowcolmatch/astropix3rowcol.cpp
#include <vector>
#include <functional>
#include <string>

struct HalfHit {
    int layer;
    int chipID;
    int location;     // row or column index
    int isCol;        // 0 = row, 1 = col
    int timestamp;
    int tot_total;
    double tot_us;
    long long fpga_ts;
};

struct MatchedHit {
    int layer;
    int chipID;
    int row;
    int col;
    int row_timestamp;
    int col_timestamp;
    int row_tot;
    int col_tot;
    double row_tot_us;
    double col_tot_us;
    long long row_fpga_ts;
    long long col_fpga_ts;
};

std::vector<MatchedHit> rowcolmatch(
    const std::vector<HalfHit>& chip0,
    std::function<bool(int,int)> fts =
        [](int x, int y) { return (x - y == 0) || (x - y == 1); },
    std::function<bool(int,int)> ftot =
        [](int x, int y) { return (x - y > 6) && (x - y < 15); }
) {
    std::vector<MatchedHit> output;

    size_t linenb = 0;
    while (linenb < chip0.size()) {
        const HalfHit& rowHit = chip0[linenb];

        // Try to match only rows
        if (rowHit.isCol == 0) {
            bool foundcol = false;
            size_t i = linenb + 1;

            while (i < chip0.size() &&
                   (!foundcol || chip0[i].isCol == 1)) {

                const HalfHit& colHit = chip0[i];

                if (colHit.isCol == 1) {
                    if (!foundcol)
                        foundcol = true;

                    if (fts(rowHit.timestamp, colHit.timestamp) &&
                        ftot(rowHit.tot_total, colHit.tot_total)) {

                        MatchedHit mh;
                        mh.layer = rowHit.layer;
                        mh.chipID = rowHit.chipID;
                        mh.row = rowHit.location;
                        mh.col = colHit.location;
                        mh.row_timestamp = rowHit.timestamp;
                        mh.col_timestamp = colHit.timestamp;
                        mh.row_tot = rowHit.tot_total;
                        mh.col_tot = colHit.tot_total;
                        mh.row_tot_us = rowHit.tot_us;
                        mh.col_tot_us = colHit.tot_us;
                        mh.row_fpga_ts = rowHit.fpga_ts;
                        mh.col_fpga_ts = colHit.fpga_ts;

                        output.push_back(mh);
                    }
                }
                ++i;
            }
        }
        ++linenb; // Skip cols
    }

    return output;
}

#include <iostream>
#include <cstdlib>

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: rowcolmatch <filename>\n";
        return 1;
    }

    int layers = 3;
    int chips = 4;
    int mints = 0;
    int maxts = 1;
    int mintot = 6;
    int maxtot = 15;

    // Assume data has already been read from CSV
    std::vector<HalfHit> data;   // decoded halfhits
    std::vector<HalfHit> dataf;  // filtered halfhits

    // Example filtering (payload & location filtering omitted for brevity)
    for (const auto& h : data) {
        if (h.location < 35) {
            dataf.push_back(h);
        }
    }

    std::vector<MatchedHit> allMatches;

    for (int layer = 0; layer < layers; ++layer) {
        for (int chip = 0; chip < chips; ++chip) {
            std::vector<HalfHit> datac;
            for (const auto& h : dataf) {
                if (h.layer == layer && h.chipID == chip) {
                    datac.push_back(h);
                }
            }

            auto matches = rowcolmatch(
                datac,
                [&](int x, int y) {
                    int d = x - y;
                    return d >= mints && d <= maxts;
                },
                [&](int x, int y) {
                    int d = x - y;
                    return d >= mintot && d <= maxtot;
                }
            );

            std::cout << "Layer " << layer
                      << ", Chip " << chip
                      << ": " << datac.size()
                      << " halfhits found, "
                      << matches.size()
                      << " hits matched\n";

            allMatches.insert(allMatches.end(),
                              matches.begin(),
                              matches.end());
        }
    }

    // Write allMatches to CSV (not shown)

    return 0;
}




